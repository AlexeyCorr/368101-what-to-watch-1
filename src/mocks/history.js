export const history = {
  action: `PUSH`,
  block: jest.fn(),
  createHref: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  length: 2,
  listen: jest.fn(),
  location: {pathname: `/login`, search: ``, hash: ``, state: undefined, key: `rja7sx`},
  push: jest.fn(),
  replace: jest.fn(),
};
