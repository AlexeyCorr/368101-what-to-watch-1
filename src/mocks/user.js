export const user = {
  id: 1,
  email: `w@main.com`,
  name: `w`,
  avatarUrl: `/wtw/static/avatar/7.jpg`
};
