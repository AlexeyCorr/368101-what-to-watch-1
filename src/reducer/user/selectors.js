import NameSpace from './../name-spaces.js';

const NAME_SPACE = NameSpace.USER;

export const getUser = (state) => {
  return state[NAME_SPACE].user;
};

export const getError = (state) => {
  return state[NAME_SPACE].error;
};
