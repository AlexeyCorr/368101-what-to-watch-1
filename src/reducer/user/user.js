import camelcaseKeys from 'camelcase-keys';

const initialState = {
  user: {},
  error: undefined,
};

const ActionType = {
  LOG_IN: `LOG_IN`,
  LOG_ERROR: `LOG_ERROR`,
};

const Operation = {
  logIn: (email, password) => (dispatch, _getState, api) => {
    return api.post(`/login`, {
      email,
      password
    }).then((response) => {
      const data = camelcaseKeys(response.data);

      dispatch(ActionCreator.logIn(data));
    }).catch((error) => {
      dispatch(ActionCreator.logError(error.message));
    });
  }
};

const ActionCreator = {
  logIn: (user) => {
    return {
      type: ActionType.LOG_IN,
      payload: user,
    };
  },
  logError: (error) => {
    return {
      type: ActionType.LOG_ERROR,
      payload: error,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOG_IN:
      return Object.assign({}, state, {
        ...state,
        user: action.payload,
        error: undefined,
      });

    case ActionType.LOG_ERROR:
      return Object.assign({}, state, {
        ...state,
        error: action.payload,
      });

    default:
      return state;
  }
};


export {
  ActionCreator,
  ActionType,
  Operation,
  reducer,
};
