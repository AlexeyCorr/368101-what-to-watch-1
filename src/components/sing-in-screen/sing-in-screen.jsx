import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {getUser, getError} from './../../reducer/user/selectors.js';
import {Operation, ActionCreator} from './../../reducer/user/user.js';
import Path from './../../paths.js';

import Header from './../header/header.jsx';
import Footer from './../footer/footer.jsx';

class SingInScreen extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      email: undefined,
      password: undefined,
    };

    this._onSubmit = this._onSubmit.bind(this);
    this._onChengeEmailInput = this._onChengeEmailInput.bind(this);
    this._onChengePasswordInput = this._onChengePasswordInput.bind(this);
  }

  render() {
    const {user, error} = this.props;

    return (
      <div className="user-page">

        <Header className={`user-page__head`} title={`Sign in`} user={user}/>

        <div className="sign-in user-page__content">
          <form
            action="#"
            className="sign-in__form"
            onSubmit={this._onSubmit}
          >
            <div className="sign-in__message">
              <p>{error}</p>
            </div>
            <div className="sign-in__fields">
              <div className="sign-in__field">
                <input
                  className="sign-in__input"
                  type="email"
                  autoComplete="username email"
                  placeholder="Email address"
                  name="user-email"
                  id="user-email"
                  onChange={this._onChengeEmailInput}
                />
                <label className="sign-in__label visually-hidden" htmlFor="user-email">Email address</label>
              </div>
              <div className="sign-in__field">
                <input
                  className="sign-in__input"
                  type="password"
                  placeholder="Password"
                  name="user-password"
                  id="user-password"
                  autoComplete="current-password"
                  onChange={this._onChengePasswordInput}
                />
                <label className="sign-in__label visually-hidden" htmlFor="user-password">Password</label>
              </div>
            </div>
            <div className="sign-in__submit">
              <button
                className="sign-in__btn"
                type="submit"
              >Sign in</button>
            </div>
          </form>
        </div>

        <Footer/>
      </div>
    );
  }

  componentDidMount() {
    const {user, history} = this.props;

    if (user.id) {
      history.push(Path.MAIN);
    }
  }

  _onSubmit(evt) {
    const {logIn, logError, history} = this.props;
    const {email, password} = this.state;

    evt.preventDefault();

    if (email && password) {
      logIn(email, password).then(() => {
        if (history.length > 1) {
          history.goBack();
        } else {
          history.push(Path.MAIN);
        }
      });
    } else {
      logError(`Fields email and password is required.`);
    }
  }

  _onChengeEmailInput(evt) {
    const {target} = evt;

    if (target) {
      this.setState({
        email: target.value
      });
    }
  }

  _onChengePasswordInput(evt) {
    const {target} = evt;

    if (target) {
      this.setState({
        password: target.value
      });
    }
  }
}

SingInScreen.propTypes = {
  user: PropTypes.object,
  logIn: PropTypes.func.isRequired,
  history: PropTypes.object,
  logError: PropTypes.func,
  error: PropTypes.string,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  user: getUser(state),
  error: getError(state),
});

const mapDispachToProps = (dispatch) => ({
  logIn: (email, password) => dispatch(Operation.logIn(email, password)),
  logError: (error) => dispatch(ActionCreator.logError(error)),
});

export {SingInScreen};

export default connect(mapStateToProps, mapDispachToProps)(SingInScreen);
