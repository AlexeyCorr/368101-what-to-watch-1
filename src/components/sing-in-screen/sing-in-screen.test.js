import React from 'react';
import renderer from 'react-test-renderer';
import {StaticRouter} from 'react-router';

import {history} from './../../mocks/history.js';
import {user} from './../../mocks/user.js';
import {SingInScreen} from './sing-in-screen.jsx';

it(`renders correctly`, () => {
  const logIn = jest.fn();
  const tree = renderer
    .create(
      <StaticRouter>
        <SingInScreen user={user} logIn={logIn} history={history}/>
      </StaticRouter>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
