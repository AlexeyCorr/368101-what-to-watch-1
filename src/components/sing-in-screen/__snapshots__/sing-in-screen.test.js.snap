// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--undefined"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      Sign in
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <a
          href="/favorites"
          onClick={[Function]}
        >
          <img
            alt="User avatar"
            height="63"
            src="https://es31-server.appspot.com/wtw/static/avatar/7.jpg"
            width="63"
          />
        </a>
      </div>
    </div>
  </header>
  <div
    className="sign-in user-page__content"
  >
    <form
      action="#"
      className="sign-in__form"
      onSubmit={[Function]}
    >
      <div
        className="sign-in__message"
      >
        <p />
      </div>
      <div
        className="sign-in__fields"
      >
        <div
          className="sign-in__field"
        >
          <input
            autoComplete="username email"
            className="sign-in__input"
            id="user-email"
            name="user-email"
            onChange={[Function]}
            placeholder="Email address"
            type="email"
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-email"
          >
            Email address
          </label>
        </div>
        <div
          className="sign-in__field"
        >
          <input
            autoComplete="current-password"
            className="sign-in__input"
            id="user-password"
            name="user-password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
          />
          <label
            className="sign-in__label visually-hidden"
            htmlFor="user-password"
          >
            Password
          </label>
        </div>
      </div>
      <div
        className="sign-in__submit"
      >
        <button
          className="sign-in__btn"
          type="submit"
        >
          Sign in
        </button>
      </div>
    </form>
  </div>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        Â© 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
